server/src/main/java/io/crate/role/Roles.java

```diff
@@ -32,26 +32,31 @@
 
 public interface Roles {
 
-    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, Boolean> HAS_PRIVILEGE_FUNCTION = (r, t, c, o) ->
-        r.privileges().matchPrivilege(t, c, (String) o);
+    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, PrivilegeState> HAS_PRIVILEGE_FUNCTION =
+        (r, t, c, o) -> r.privileges().matchPrivilege(t, c, (String) o);
 
-    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, Boolean> HAS_ANY_PRIVILEGE_FUNCTION = (r, t, c, o) ->
-        r.privileges().matchPrivilegeOfAnyType(c, (String) o);
+    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, PrivilegeState> HAS_ANY_PRIVILEGE_FUNCTION =
+        (r, t, c, o) -> r.privileges().matchPrivilegeOfAnyType(c, (String) o);
 
-    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, Boolean> HAS_SCHEMA_PRIVILEGE_FUNCTION =
+    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, PrivilegeState> HAS_SCHEMA_PRIVILEGE_FUNCTION =
         (r, t, c, o) -> {
+            PrivilegeState result = PrivilegeState.REVOKE;
             for (Privilege privilege : r.privileges()) {
-                if (privilege.state() == PrivilegeState.GRANT && privilege.ident().type() == t) {
-                    if (privilege.ident().clazz() == Privilege.Clazz.CLUSTER) {
-                        return true;
-                    }
-                    if (privilege.ident().clazz() == Privilege.Clazz.SCHEMA &&
-                        OidHash.schemaOid(privilege.ident().ident()) == (Integer) o) {
-                        return true;
+                if (privilege.ident().type() == t) {
+                    if (privilege.ident().clazz() == Privilege.Clazz.SCHEMA
+                        && OidHash.schemaOid(privilege.ident().ident()) == (Integer) o) {
+
+                        if (privilege.state() == PrivilegeState.DENY || privilege.state() == PrivilegeState.GRANT) {
+                            return privilege.state();
+                        }
```

Comment:
Why can't we just return the state in any case? Revoking a privileges will remove any matching privilege...

Reply:
You're right we can return here directly.